### PR TITLE
[SYCL][fp64][SYCLLowerIR] Fix an issue when double type inside multiple nested structs is not identified

### PR DESCRIPTION
**Pull request by asudarsa**
There are cases where some instructions use data structures that contain double data type deeply nested inside multiple levels.
Example:
struct T1 {
  double x;
};
struct T2 {
  struct T1 y;
};
T2 x;

In such case, the fp64 emulation support in SYCLPropagateAspectUsage pass was not correctly identifying T2 as a type that has doubles. This led to wrong behavior when adding fp64 aspect.
This PR fixes the issue.

Thanks

### CODE REVIEW

**Comment by asudarsa** [comment]
@AlexeySachkov 

Can you please take a look? 

Thanks

File: llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp [hunk 1]
```diff
@@ -267,7 +267,7 @@ bool hasDoubleType(const Type *T) {
   if (T->isDoubleTy())
     return true;
   for (const Type *TT : T->subtypes())
-    if (TT->isDoubleTy())
+    if (hasDoubleType(TT))
```
Comment (AlexeySachkov) [inline llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp:270]: I'm concerned about potential infinite recursion here. There is a similar function which collects aspects used by a type and it uses a cache to ensure that no type is visited twice.

I would recommend checking an example where you have something like this:
```c++
struct B;
struct A {
 B *b;
};
struct B {
  A *a;
};
```

Comment (asudarsa) [reply llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp:270]: Hi @AlexeySachkov 

Thanks. That was a good catch.
I tried this locally and LLVM IR generated looks like this:
%struct.T1 = type { ptr addrspace(4) }
%struct.T2 = type { ptr addrspace(4) }
So, we are saved here by opaque pointer types.
I added a test for this
Thanks

**Review by steffenlarsen — APPROVED** [review]
🚀

**Comment by asudarsa** [comment]
@intel/llvm-gatekeepers 

Think this is ready to be merged.

Thanks

**Pull request merged**